### PR TITLE
Introduce BoolVar struct

### PR DESCRIPTION
**Pull request by alxiong**
## Description

Introducing `struct BoolVar` whose creation automatically enforces a `bool_gate()`, distinguishing from normal `Variable` to avoid repeated checks and make API clearer.

closes: #91 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] ~Wrote unit tests~
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Review by chancharles92 — CHANGES_REQUESTED** [review]
@alxiong Nice work. There is one subtlety that needs to be addressed though.

File: plonk/src/circuit/customized/ecc/mod.rs
Comment (alxiong) [reply plonk/src/circuit/customized/ecc/mod.rs:854]: I'm not sure I fully sympathize with the concern. 
these variables are created by circuit dev, he/she can name it however they want, all `create_bool_variable()` does is making sure the created variable has an underlying value of either `F::zero()` or `F::one()`.

I currently didn't offer API for _constant_ true/false var, which is maybe your `circuit.false()` refers to.

Comment (chancharles92) [reply plonk/src/circuit/customized/ecc/mod.rs:854]: Imagine a naive circuit that returns true only if the secret witness of `var` is false. Suppose the circuit is created as follows:
`let false_var = circuit.create_bool_variable(false);`
`let var = circuit.create_variable(F::zero);`
`circuit.equal_gate(false_var, var);`
This circuit will return true no matter whether `wit[var] = wit[false_var] = true` or `wit[var] = wit[false_var] = false`. The right way to build this circuit is
`let false_var = circuit.zero()`
`let var = circuit.create_variable(F::zero);`
`circuit.equal_gate(false_var, var);`.

Comment (alxiong) [reply plonk/src/circuit/customized/ecc/mod.rs:854]: sry, I still don't find your example problematic. maybe I miss something.

> This circuit will return true no matter whether wit[var] = wit[false_var] = true or wit[var] = wit[false_var] = false. 

1. `wit[false_var]` was set by the circuit dev (during `create_bool_variable(false)`), not by malicious users who try to provide (instance, witness) to derive a satisfying assignment. 
2. if the circuit dev accidentally change `wit[false_var]`, by the same fashion, he can change `wit[0]` and change the value of `circuit.zero()`

That said, I do see one single difference in the two implementations above:
the latter uses 1 less variable 

iiuc, the analogy of the two approaches is: 
```rust
fn is_positive_gate(cs: &mut Circuit, a: Variable) {
  let zero_var = cs.create_variable(F::zero()); // create a new variable, but of fixed value
  cs.less_than_gate(a, zero_var);
}

fn is_positive_gate(cs: &mut Circuit, a: Variable) {
  let zero_var = cs.zero(); // utilize a existing value of fixed value that I already know
  cs.less_than_gate(a, zero_var);
}
```
both are correct imo.

Comment (alxiong) [reply plonk/src/circuit/customized/ecc/mod.rs:854]: we can give circuit dev the option:

```rust
    /// Return a default variable with value `false` (namely zero).
    fn false_var(&self) -> BoolVar {
        BoolVar::new_unchecked(self.zero())
    }

    /// Return a default variable with value `true` (namely one).
    fn true_var(&self) -> BoolVar {
        BoolVar::new_unchecked(self.one())
    }
```

File: plonk/src/circuit/customized/ecc/mod.rs
Comment (alxiong) [reply plonk/src/circuit/customized/ecc/mod.rs]: update in [6a19265](https://github.com/EspressoSystems/jellyfish/pull/94/commits/6a1926505fbfa3f2b0e065afe3d853708d9c997c) see if you like it.

File: plonk/src/circuit/customized/ecc/glv.rs
Comment (zhenfeizhang) [inline plonk/src/circuit/customized/ecc/glv.rs:518]: looks like `k2_is_pos_sat` is an output of `circuit.check_equal(left_var, right_var)?` which makes me think that `circuit.check_equal(left_var, right_var)?` should output a BoolVar instead of a `Variable`. WDYT?

**Comment by zhenfeizhang** [comment]
In general I want to advocate my "circuit ref" idea. For example 
``` rust
struct BoolVar<'a> {
  index: Variable,
  cs: &'a ConstraintSystem
}
```

On top of overloading operators here are some additional thoughts:
- to access the variable, we should use an API `get_index(&self)->Variable` -- i think it is a bit cleaner than `into()`
  - it makes more sense to me to implement Into/From for `Vec<BoolVar>` and `Variable`, in both directions, which are binary (de)composition
- to create new variables we should use associated function `BoolVar::new(cs: &ConstraintSystem, value: bool)` rather than `Circuit::create_bool_variable(...)` and so on... 
- to check variable bound we can simply call `bool_var.check_bound()`, and we do not need to pass parameters

File: plonk/src/circuit/basic.rs
Comment (zhenfeizhang) [inline plonk/src/circuit/basic.rs:272]: slightly prefer `boolean_variable` over `bool_variable`

File: plonk/src/circuit/customized/ecc/mod.rs
Comment (chancharles92) [inline plonk/src/circuit/customized/ecc/mod.rs:275]: Can we replace `self.mul(x_eq, y_eq)` (in the olde commit) with `self.logic_and(x_eq: BoolVar, y_eq : BoolVar) -> BoolVar`? We can simplify the code here by doing that.

File: plonk/src/circuit/customized/ecc/glv.rs
Comment (alxiong) [reply plonk/src/circuit/customized/ecc/glv.rs:518]: yeah exactly, currently we have: `pub fn check_equal(&mut self, a: Variable, b: Variable) -> Result<BoolVar, PlonkError> ` which is why you see here `k2_is_pos_sat.into()` since `k2_is_pos_sat` is `BoolVar` type.

File: plonk/src/circuit/customized/ecc/mod.rs
Comment (alxiong) [reply plonk/src/circuit/customized/ecc/mod.rs:275]: oh yes, good suggestion

File: plonk/src/circuit/customized/mod.rs
Comment (chancharles92) [reply plonk/src/circuit/customized/mod.rs]: oh i see, right

File: plonk/src/circuit/customized/ecc/mod.rs
Comment (alxiong) [reply plonk/src/circuit/customized/ecc/mod.rs]: done in [5f0cd66](https://github.com/EspressoSystems/jellyfish/pull/94/commits/5f0cd663417b46145cedbd3f620aa8963808ba00)

File: plonk/src/circuit/basic.rs
Comment (alxiong) [reply plonk/src/circuit/basic.rs]: done in [5f0cd66](https://github.com/EspressoSystems/jellyfish/pull/94/commits/5f0cd663417b46145cedbd3f620aa8963808ba00)

**Review by chancharles92 — APPROVED** [review]
LGTM, we can wait for Zhenfei's approval before merging.

File: plonk/src/circuit/customized/ecc/glv.rs
Comment (zhenfeizhang) [reply plonk/src/circuit/customized/ecc/glv.rs:518]: cool. Sorry I somehow thought the `into()` here was the reversed direction

**Comment by zhenfeizhang** [comment]
LGTM2

**Pull request merged**
